File: sdk/test/sessionId.test.ts
import { describe, expect, it } from 'vitest';
import { deriveSessionId, sessionConfigPhysicalHash } from '../src/sessionId.js';

const cfg = {
  sessionKey: '0x0000000000000000000000000000000000000002',
  token: '0x00000000000000000000000000000000000000aa',
  allowedTarget: '0x00000000000000000000000000000000000000aa',
  allowedRecipient: '0x0000000000000000000000000000000000000b01',
  allowedSelector: '0xa9059cbb',
  maxPerTransfer: 100_000_000n,
  maxTotalSpend: 500_000_000n,
  validAfter: 1n,
  validUntil: 9_999_999_999n,
  allowBatch: false,
} as const;

describe('sessionId', () => {
  it('derives deterministic session id', () => {
    const physical = sessionConfigPhysicalHash(cfg);
    expect(physical).toMatch(/^0x[a-f0-9]{64}$/);
    const id = deriveSessionId({
      chainId: 84532,
      adapter: '0x00000000000000000000000000000000000000c0',
      account: '0x0000000000000000000000000000000000000001',
      nonceKey: 12345n,
      config: cfg,
    });
    expect(id).toBe('0x52a14e7814be7dbf606ee36eb57bef03d9d9e50b72bd13097f14eb123d26b936');
  });
});

File: signgate/internal/session/sessionid_test.go
package session

import "testing"

func TestDeriveSessionIDDeterministic(t *testing.T) {
	cfg := Config{
		Account:          "0x0000000000000000000000000000000000000001",
		NonceKey:         "12345",
		SessionKey:       "0x0000000000000000000000000000000000000002",
		Token:            "0x00000000000000000000000000000000000000aa",
		AllowedTarget:    "0x00000000000000000000000000000000000000aa",
		AllowedRecipient: "0x0000000000000000000000000000000000000b01",
		AllowedSelector:  "0xa9059cbb",
		MaxPerTransfer:   "100000000",
		MaxTotalSpend:    "500000000",
		ValidAfter:       1,
		ValidUntil:       9999999999,
		AllowBatch:       false,
		PolicyHash:       "0x0000000000000000000000000000000000000000000000000000000000000011",
		ChainID:          84532,
		AdapterAddress:   "0x00000000000000000000000000000000000000c0",
	}
	if err := ValidateV1(cfg); err != nil {
		t.Fatal(err)
	}
	a, err := DeriveSessionID(cfg)
	if err != nil {
		t.Fatal(err)
	}
	b, err := DeriveSessionID(cfg)
	if err != nil {
		t.Fatal(err)
	}
	if a != b {
		t.Fatalf("session id not deterministic")
	}
}
